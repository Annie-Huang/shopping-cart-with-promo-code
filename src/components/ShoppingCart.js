import React, {useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import sumby from 'lodash.sumby';
import * as discountCodeActions from "../redux/actions/discountCodeActions";
import ShoppingCartService from '../lib/ShoppingCartService';

export const ShoppingCart  = ({
    discountCodes,
    loadDiscountCodes,
    cartItems,
}) => {
    const [total, setTotal] = useState(0);
    const [promoCode, setPromoCode] = useState('');
    const [discountAmount, setDiscountAmount] = useState(0);


    useEffect(() => {
        loadDiscountCodes().catch(error => {
            alert("Loading discount codes failed" + error);
        });
    }, [loadDiscountCodes]);

    useEffect(() => {
        setTotal(Number(sumby(cartItems, 'subTotal').toFixed(2)));
    }, [cartItems]);

    const discountCodeList = discountCodes.map(discountCode =>
        <div key={discountCode.id}>{discountCode.id} : {discountCode.description}</div>
    );

    const cartItemList = cartItems.map((cartItem, index) =>
        <div className='row' key={index}>
            <div className='col-md-2'>{cartItem.product.name}</div>
            <div className='col-md-2'>Unit Price: ${cartItem.product.price}</div>
            <div className='col-md-2'>Quantity: {cartItem.quantity}</div>
            <div className='col-md-2'>Subtotal: ${cartItem.subTotal}</div>
        </div>
    );

    const applyPromoCode = () => {
        console.log('promoCode=', promoCode);
        setDiscountAmount(
            promoCode === '' ? 0 : ShoppingCartService.calculateDiscountAmount(cartItems, total, promoCode)
        );
    };

    return (
        <div>
            <h5>Available discount codes are as following. Please keep in mind that you can only apply for one.</h5>
            {discountCodeList}
            {cartItems.length > 0 &&
            // {1==1 &&
            <div className="card">
                <div className="card-header bg-info text-white">
                    <h3 className="card-title">Your Basket:</h3>
                </div>
                <div className="card-body">
                    {cartItemList}
                    <br/>
                    <div className='row'>
                        <div className='col-md-12'>
                            <h3 className="card-title">SubTotal: ${total}</h3>
                        </div>
                    </div>
                </div>
                <div className="card-footer">
                    <div className='row'>
                        <div className='col-sm-1'>Discount Code:</div>
                        <input className='col-sm-1' type="text"  value={promoCode} onChange={e => setPromoCode(e.target.value)}/>
                        <div className='col-sm-1'>
                            <button type="button"
                                    className="btn btn-sm btn-primary"
                                    onClick={() => applyPromoCode()}
                            >Apply</button>
                        </div>
                    </div>
                    <br/>
                    <div className='row'>
                        <div className='col-sm-12'>You have save: ${discountAmount}</div>
                    </div>
                    <br/>
                    <div className='row'>
                        <div className='col-md-12'>
                            <h3 className="card-title">Total: ${total-discountAmount}</h3>
                        </div>
                    </div>
                    <div className='row'>
                        <div className='col-lg-2'>
                        <button type="button" className="btn btn-lg btn-primary">Checkout</button>
                        </div>
                    </div>
                </div>
            </div>}
        </div>
    );
};

ShoppingCart.propTypes = {
    discountCodes: PropTypes.array.isRequired,
    cartItems: PropTypes.array.isRequired,
    loadDiscountCodes: PropTypes.func.isRequired
};


const mapStateToProps = (state, ownProps) => ({
    discountCodes: state.discountCodes,
    cartItems: state.cartItems,
});

const mapDispatchToProps = {
    loadDiscountCodes: discountCodeActions.loadDiscountCodes
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ShoppingCart);
