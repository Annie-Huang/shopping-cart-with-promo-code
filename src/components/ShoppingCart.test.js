import {shallow, mount} from "enzyme/build";
import {ShoppingCart} from "./ShoppingCart";
import React from "react";
import toJson from "enzyme-to-json";


describe('ShoppingCart presentation', () => {
    const clearCart = jest.fn();
    const cartItems = require('../resources/fixtures/cart-items.json');

    const props = {
        cartItems,
        clearCart
    };

    it('should render ShoppingCart component according to the cartItems information', () => {
        expect(toJson(shallow(<ShoppingCart {...props} />))).toMatchSnapshot();
    });

    it('#clearCart should be called when clicking Clear Cart button', () => {
        const wrapper = shallow(<ShoppingCart {...props} />);

        const buttons = wrapper.find('button');
        const clearButton =  buttons.first();
        expect(clearButton.text()).toBe('Clear Cart');

        clearButton.simulate('click');
        expect(clearCart).toBeCalled();
    });

    it('#applyPromoCode should be called when clicking Apply Code button. And if promoCode is valid, then discount amount, total should be recalculated', () => {
        const wrapper = mount(<ShoppingCart {...props} />);
        wrapper.find('input').simulate('change', { target: { value: 'RRD4D32' } });

        const buttons = wrapper.find('button');
        const applyCodeButton =  buttons.at(1);
        expect(applyCodeButton.text()).toBe('Apply Code');

        applyCodeButton.simulate('click');
        // 199.99 + 1998 = 2197.99
        expect(wrapper.find('#subTotal').text()).toEqual('SubTotal: $2,197.99');
        expect(wrapper.find('#discount').text()).toEqual('You have save: $219.80');
        expect(wrapper.find('#total').text()).toEqual('Total: $1,978.19');
    });

    it('#applyPromoCode should be called when clicking Apply Code button. And if promoCode is invalid, then discount amount remain 0', () => {
        const wrapper = mount(<ShoppingCart {...props} />);
        wrapper.find('input').simulate('change', { target: { value: 'ABC' } });

        const buttons = wrapper.find('button');
        const applyCodeButton =  buttons.at(1);
        expect(applyCodeButton.text()).toBe('Apply Code');

        applyCodeButton.simulate('click');
        // 199.99 + 1998 = 2197.99
        expect(wrapper.find('#subTotal').text()).toEqual('SubTotal: $2,197.99');
        expect(wrapper.find('#discount').text()).toEqual('You have save: $0.00');
        expect(wrapper.find('#total').text()).toEqual('Total: $2,197.99');
    });

    it('#clearPromoCode should be called when clicking Clear Code button. And if promoCode input field will be clear and discount amount remain 0', () => {
        const wrapper = mount(<ShoppingCart {...props} />);
        wrapper.find('input').simulate('change', { target: { value: 'RRD4D32' } });

        const buttons = wrapper.find('button');
        const applyCodeButton =  buttons.at(1);
        expect(applyCodeButton.text()).toBe('Apply Code');
        applyCodeButton.simulate('click');
        const clearCodeButton =  buttons.at(2);
        expect(clearCodeButton.text()).toBe('Clear Code');
        clearCodeButton.simulate('click');

        expect(wrapper.find('input').instance().value).toEqual('');
        // 199.99 + 1998 = 2197.99
        expect(wrapper.find('#subTotal').text()).toEqual('SubTotal: $2,197.99');
        expect(wrapper.find('#discount').text()).toEqual('You have save: $0.00');
        expect(wrapper.find('#total').text()).toEqual('Total: $2,197.99');
    });

    it('#update cartItems will trigger discount amount, total should be recalculated', () => {
        const wrapper = mount(<ShoppingCart {...props} />);
        wrapper.find('input').simulate('change', { target: { value: 'RRD4D32' } });

        const buttons = wrapper.find('button');
        const applyCodeButton =  buttons.at(1);
        expect(applyCodeButton.text()).toBe('Apply Code');

        applyCodeButton.simulate('click');
        // 199.99 + 1998 = 2197.99
        expect(wrapper.find('#subTotal').text()).toEqual('SubTotal: $2,197.99');
        expect(wrapper.find('#discount').text()).toEqual('You have save: $219.80');
        expect(wrapper.find('#total').text()).toEqual('Total: $1,978.19');


        // What we really want to test is from this section onwards...
        const newCartItems = cartItems.slice(0);
        newCartItems[0].quantity = 2;
        newCartItems[0].subTotal = 399.98;
        wrapper.setProps({ cartItems: newCartItems });

        // // 399.98 + 1998 = 2397.98
        expect(wrapper.find('#subTotal').text()).toEqual('SubTotal: $2,397.98');
        expect(wrapper.find('#discount').text()).toEqual('You have save: $239.80');
        expect(wrapper.find('#total').text()).toEqual('Total: $2,158.18');
    });
});
