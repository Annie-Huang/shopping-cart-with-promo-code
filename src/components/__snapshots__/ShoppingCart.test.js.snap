// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCart presentation should render ShoppingCart component according to the cartItems information 1`] = `
<Fragment>
  <div
    className="card"
  >
    <div
      className="card-header bg-info text-white"
    >
      <h3
        className="card-title"
      >
        Your Basket:
      </h3>
    </div>
    <div
      className="card-body"
    >
      <div
        className="row"
        key="0"
      >
        <div
          className="col-md-2"
        >
          Workflow
        </div>
        <div
          className="col-md-2"
        >
          Unit Price: $
          199.99
        </div>
        <div
          className="col-md-2"
        >
          Quantity: 
          1
        </div>
        <div
          className="col-md-2"
        >
          Subtotal: $
          199.99
        </div>
      </div>
      <div
        className="row"
        key="1"
      >
        <div
          className="col-md-2"
        >
          Document Generation
        </div>
        <div
          className="col-md-2"
        >
          Unit Price: $
          9.99
        </div>
        <div
          className="col-md-2"
        >
          Quantity: 
          200
        </div>
        <div
          className="col-md-2"
        >
          Subtotal: $
          1998
        </div>
      </div>
      <br />
      <div
        className="row"
      >
        <div
          className="col-md-2"
        >
          <h3
            className="card-title"
            id="subTotal"
          >
            SubTotal: $
            0
          </h3>
        </div>
        <div
          className="col-md-2"
        >
          <button
            className="btn btn-sm btn-primary"
            onClick={[Function]}
            type="button"
          >
            Clear Cart
          </button>
        </div>
      </div>
    </div>
    <div
      className="card-footer"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-2"
        >
          Discount Code:
        </div>
        <input
          className="col-sm-1"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="col-sm-4"
        >
          <button
            className="btn btn-sm btn-primary"
            onClick={[Function]}
            type="button"
          >
            Apply Code
          </button>
            
          <button
            className="btn btn-sm btn-secondary"
            onClick={[Function]}
            type="button"
          >
            Clear Code
          </button>
        </div>
      </div>
      <br />
      <div
        className="row"
      >
        <div
          className="col-sm-12"
          id="discount"
        >
          You have save: $
          0
        </div>
      </div>
      <br />
      <div
        className="row"
      >
        <div
          className="col-md-12"
        >
          <h3
            className="card-title"
            id="total"
          >
            Total: $
            0
          </h3>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-lg-2"
        >
          <button
            className="btn btn-lg btn-primary"
            type="button"
          >
            Checkout
          </button>
        </div>
      </div>
    </div>
  </div>
</Fragment>
`;
